File: BrokenLCD/cpp/blcd.cpp
// Broken LCD cpp soln for Code Eval by Steven A Dunn

#include <cstdlib>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using std::cout;
using std::endl;
using std::ifstream;
using std::istringstream;
using std::string;
using std::to_string;
using std::vector;

// each display from 0 to 9, plus the decimal point, stored as decimals
int numbers[] = {252, 96, 218, 242, 102, 182, 190, 224, 254, 246, 1};

vector<string> tokenize(string, char);
string& rtrim(string&);

int main (int argc, char* const argv[])
{
    ifstream input_file(argv[1]);
    string line;

    if (input_file)
    {
        while (getline(input_file, line))
        {
            vector<string> tokens = tokenize(line, ';');
            vector<string> displays = tokenize(tokens[0], ' ');
            string number = tokens[1];

            vector<int> converted_numbers;
            for (int i = 0; i < number.size(); ++i)
            {
                char* c = &number[i];
                if (c[0] != '.')
                {
                    int x = number[i] - '0';
                    converted_numbers.push_back(numbers[x]);
                }
                else
                    converted_numbers[i-1] += 1;
            }

            bool did_print = false;
            for (int i = 0; i < displays.size() - converted_numbers.size() + 1; ++i)
            {
                int match = 0;
                for (int j = 0; j < converted_numbers.size(); ++j)
                {
                    char *str;
                    int converted_display = strtol (displays[i+j].c_str(), &str, 2);
                    if ((converted_display & converted_numbers[j]) != converted_numbers[j])
                        break;
                    else
                        match += 1;
                }
                if (match == converted_numbers.size())
                {
                    puts("1");
                    did_print = true;
                    break;
                }
            }
            if (!did_print)
                puts("0");

        }
        input_file.close();
    }
    return 0;
}

vector<string> tokenize(string input, char delim)
{
    vector<string> tokens;
    istringstream iss(input);
    string line;
    while (getline(iss, line, delim))
        tokens.push_back(line);
    return tokens;
}

File: BrokenLCD/py2/blcd.py
# broken lcd soln in py2 for code eval by steven a dunn

import sys

# each display from 0 to 9, plus the decimal point
numbers = [0b11111100, 0b01100000, 0b11011010, 0b11110010, 0b01100110, 0b10110110, 0b10111110, 0b11100000, 0b11111110, 0b11110110, 0b00000001]

f = open(sys.argv[1], 'r')
for line in f:
    displays, number = line.split(";")
    displays = displays.split(" ")
    number = number.strip()

    converted_numbers = []
    for i in range(0, len(number)):
        if number[i] != ".":
            converted_numbers.append(numbers[int(number[i])])
        if number[i] == ".":
            converted_numbers[i-1] += 0b00000001

    did_print = False
    print "displays size: ", len(displays), "converted_numbers size: ", len(converted_numbers)
    for i in range(0, len(displays) - len(converted_numbers) + 1):
        match = 0
        for j in range(0, len(converted_numbers)):
            print "i: ", i, "j: ", j
            print displays[i+j];
            print int(displays[i+j], 2)
            print converted_numbers[j]
            print int(displays[i+j], 2) & converted_numbers[j]
            if int(displays[i+j], 2) & converted_numbers[j] != converted_numbers[j]:
                break
            else:
                match += 1
        if match == len(converted_numbers):
            print 1
            did_print = True
            break
    if not did_print:
        print 0
f.close()